### PR TITLE
solution

### PR DESCRIPTION
**Pull request by tiserett**
[DEMO LINK](https://tiserett.github.io/react_static-list-of-posts/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -2,108 +2,32 @@ import React from 'react';
...
+    comment.postId === postId
+  ));
 
-            <a className="UserInfo" href="mailto:Julianne.OConner@kory.org">
-              Patricia Lebsack
-            </a>
-          </p>
-        </div>
+  return comments.length > 0 ? comments : null;
+}
```
Comment (sergii-nosachenko) [inline src/App.tsx:19]:
```suggestion
function getComments(postId: number): Comment[] {
  return commentsFromServer.filter(comment => (
    comment.postId === postId
  ));
}
```

I have doubts about this function work. You've decided to return null if no posts found. But this decision created a new type checking branch. What if I propose you to return only one type of data - `Comment[]`. Array filter method in any case will return you an array, so you don't need to worry about data type change.

So you can remove null from everywhere and check only if `comments.length` convertes to `true`.

What do you think?

File: src/components/PostInfo/PostInfo.tsx [hunk 2]
```diff
@@ -1,5 +1,42 @@
...
+      <div className="PostInfo__header">
+        <h3 className="PostInfo__title">{ title }</h3>
+
+        <p>
+          {' Posted by  '}
+
+          {user
+            ? <UserInfo user={user} />
+            : ''}
+        </p>
```
Comment (sergii-nosachenko) [inline src/components/PostInfo/PostInfo.tsx:30]:
```suggestion
{user && <UserInfo user={user} />}
```

1. This paragraph belongs to UserInfo (if no user - no paragraph with text Posted by...)
2. Same condition can be achieved with AND operator.

File: src/components/PostInfo/PostInfo.tsx [hunk 3]
```diff
@@ -1,5 +1,42 @@
...
+      </div>
+
+      <p className="PostInfo__body">
+        { body }
+      </p>
+
+      {comments
```
Comment (sergii-nosachenko) [inline src/components/PostInfo/PostInfo.tsx:37]:
```suggestion
      {comments.length
```

If you return `Comment[]`

File: src/types/Post.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { User } from './User';
+import { Comment } from './Comment';
+
+export interface Post {
+  userId: number,
+  id: number,
+  title: string,
+  body: string,
+  comments: Comment[] | null,
```
Comment (sergii-nosachenko) [inline src/types/Post.ts:9]:
```suggestion
  comments: Comment[],
```

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]
I have found some controversial solutions it your code, that need to be discussed.

Also I've noticed that you have left styles untouched, but you need to separate'em as you did it it with components (cause they are separate BEM blocks).

File: src/App.tsx (hunk 1 above)
Comment (tiserett) [reply src/App.tsx:19]: sounds like a way to get rid of unnecessary written code

**Review by sergii-nosachenko — APPROVED** [review]
Perfect 🔥

File: src/App.tsx [hunk 5]
```diff
@@ -2,108 +2,30 @@ import React from 'react';
...
+  return commentsFromServer.filter(comment => (
+    comment.postId === postId
+  ));
+}
+
+const posts: Post[] = postsFromServer.map(post => ({
+  ...post,
+  comments: getComments(post.id),
+  user: usersFromServer.find(user => user.id === post.userId) || null,
+}));
```
Comment (AlinaMatuschak) [inline src/App.tsx:23]: you can create folder `helpers` with file `generatePosts` with this function

File: src/components/PostInfo/PostInfo.tsx [hunk 6]
```diff
@@ -1,5 +1,36 @@
...
+        {user && <UserInfo user={user} />}
+      </div>
+
+      <p className="PostInfo__body">
+        { body }
+      </p>
+
+      {comments.length
+        ? <CommentList comments={comments} />
+        : <b data-cy="NoCommentsMessage">No comments yet</b>}
```
Comment (AlinaMatuschak) [inline src/components/PostInfo/PostInfo.tsx:33]: ❤️

**Review by AlinaMatuschak — APPROVED** [review]

**Pull request closed without merging**